Step3 : 자동차 경주

hbm0411 opened the pull request:
Step3 자동차 경주 코드 구현하였습니다.
private method가 많아 테스트가 힘들어서 고민이되네요.
좋은 방법이 있으면 알려주시면 감사하겠습니다.(__)

hbm0411 committed:
feat: Readme에 자동차 경주 관련 구현사항 추가

hbm0411 committed:
feat: 사용자에게 입력 받는 UI View 구현 및 테스트코드 작성

hbm0411 committed:
feat: car model 추가

hbm0411 committed:
feat: 시도 횟수만큼 자동차 경주 구현 및 테스트 코드 작성

hbm0411 committed:
feat: 결과 출력하는 View 구현

hbm0411 commented on src/main/kotlin/raicing/processor/RaceProcessor.kt:
```diff
@@ -0,0 +1,54 @@
+        val raceResult = carList.map { car ->
+            goOrStop(car)
+
+            val afterCar = car.copy()
```
copy를 하지 않으면 리스트에 같은 객체가 들어가서 값이 계속 변하더라구요.
copy가 아닌 다른 방법이 있을까요?

laco-dev replied:
```kotlin
val carList = initCarList(carCnt)
val raceResultList = mutableListOf<RaceResult>()

repeat(raceCnt) {
     val raceResult = race(carList)
     raceResultList.add(raceResult)
}
```
앞선 코드에서 carList라는 컬렉션을 만들어서 할당한 뒤 계속해서 동일한 carList를 넣어주고 있네요.
가변적인 형태로 객체를 만들 경우 쉽게 발생하는 문제점입니다.

앞선 구조들을 개선하면서 이 부분도 같이 해결을 할 수 있을 것 같습니다.

hbm0411 replied:
Car에 deepCopy() 메소드를 만들어 수정하였습니다.
더 좋은 방법이 있다면 알려주세요 :)

hbm0411 committed:
fix: Readme의 중복 항목 삭제

laco-dev commented on src/main/kotlin/raicing/Const.kt:
```diff
@@ -0,0 +1,13 @@
+    }
+
+    object ErrorMsg {
+        const val INPUT_IS_EMPTY_OR_BLANK_ERROR_MSG = "비어있는 값은 입력 할 수 없습니다."
```
마이너한 내용입니다.
EMPTY <= BLANK 개념이라 하나만 표현해도 되지 않을까 싶네요?

hbm0411 replied:
사용하지 않는 값이라 삭제하였습니다.

laco-dev commented on src/main/kotlin/raicing/view/InputView.kt:
```diff
@@ -0,0 +1,16 @@
+import raicing.Const
+
+object InputView {
+    fun readStrAndConvertToInt(infoMsg: String): Int {
+        println(infoMsg)
+        return requireNaturalNum(readlnOrNull())
+    }
+
+    private fun requireNaturalNum(inputStr: String?): Int {
+        val inputInt = requireNotNull(inputStr?.toIntOrNull()) { Const.ErrorMsg.INPUT_IS_NOT_NATURAL_NUMBER_ERROR_MSG }
+        require(inputInt > 0) { Const.ErrorMsg.INPUT_IS_NOT_NATURAL_NUMBER_ERROR_MSG }
+        return inputInt
+    }
```
`readStrAndConvertToIn(...)` 함수가 결국엔 Int 타입을 반환하고 있어요.
이 값을 그대로 사용하게 될 도메인에서는 이 값이 0보다 클 것이라는 것을 신뢰할 수 있을까요???
0보다 큰 숫자라는 역할을 가지는 `CarCount`와 `RaceCount`라는 객체를 만들어 보면 어떨까요?

관련해서 입력값을 테스트하는 개념이 아닌 앞선 모델을 테스트 하는 형태로 변경할 수 있습니다!

laco-dev replied:
UI를 도메인 기능보다 먼저 구현 해 주셨는데,
만약 입력값을 통해 어떠한 도메인 모델을 반환하도록 기능이 구현된다면 작업했던 UI코드도 함께 수정이 불가피합니다.

UI 작업은 충분한 도메인 설계와 구현 이후 작업하면 좋아요. 🤔

laco-dev commented on src/main/kotlin/raicing/processor/RaceProcessor.kt:
```diff
@@ -0,0 +1,54 @@
+    companion object {
+        const val MIN_RANDOM_BOUND = 1
+        const val MAX_RANDOM_BOUND = 10
+        const val BASE_NUM = 4
```
상수에 의미 부여 👍

laco-dev commented on src/main/kotlin/raicing/model/RaceResult.kt:
```diff
@@ -0,0 +1,6 @@
+package raicing.model
+
+data class RaceResult(
+    val carList: List<Car>
+)
```
경주의 결과를 별도 클래스로 분리 👍

laco-dev replied:
셀 수 있는 명사의 경우는 보통 list를 붙이기 보다는 복수 형태로 만들어 주고는 합니다.
```kotlin
val cars: List<Car>
```

laco-dev replied:
파일의 마지막 라인이 두 줄이 개행되어 있어서 린트 오류가 발생하고 있네요! 🤔

laco-dev commented on src/main/kotlin/raicing/processor/RaceProcessor.kt:
```diff
@@ -0,0 +1,54 @@
+import raicing.model.RaceResult
+
+class RaceProcessor {
+    fun raceStart(carCnt: Int, raceCnt: Int): List<RaceResult> {
+
+        val carList = initCarList(carCnt)
+        val raceResultList = mutableListOf<RaceResult>()
+
+        repeat(raceCnt) {
+            val raceResult = race(carList)
+            raceResultList.add(raceResult)
+        }
+
+        return raceResultList
+    }
```
이 함수는 현재 입력받은 차량의 대수만큼 차를 생성하는 것과 경주를 하는 것 두 가지 이상의 역할을 하고 있습니다.

하나의 역할만 하도록 분리하려면 어떻게 할 수 있을까요?

hbm0411 replied:
메소드를 분리하여 main 함수에서 호출하도록 수정하였습니다.

laco-dev commented on src/main/kotlin/raicing/processor/RaceProcessor.kt:
```diff
@@ -0,0 +1,54 @@
+    }
+
+    private fun initCarList(carCnt: Int): List<Car> {
+        val carList = mutableListOf<Car>()
+        repeat(carCnt) {
+            carList.add(Car(it))
+        }
```
코틀린에서는 컬렉션을 생성하기 위한 좋은 빌더 함수들을 제공하고 있습니다.
위 코드는 아래와 같이 사용할 수 있어요.
```kotlin
val carList = List(carCnt) { Car(it) } 
```

hbm0411 replied:
와 좋은 방법이네요! 감사합니다.

laco-dev commented on src/main/kotlin/raicing/processor/RaceProcessor.kt:
```diff
@@ -0,0 +1,54 @@
+        return raceResultList
+    }
+
+    private fun initCarList(carCnt: Int): List<Car> {
```
Cnt라는 축약 표현보다는 명확하게 count로 표현 해 보면 어떨까요?
다른 개발자 누가 보더라도 어떤 변수구나를 단번에 이해할 수 있으면 좋겠어요.

laco-dev commented on README.md:
```diff
@@ -66,7 +66,7 @@
 [x] 입력받는 UI View 구현
 [] 결과를 출력하는 UI View 구현
-[] 자동차 model 구현 
+[x] 자동차 model 구현
```
단순하게 "자동차 model 구현"이라는 의미 보다는 특정 기능을 만들어 보는데 집중 해 보면 어떨까요??
관련해서 기본적인 기능에 대한 테스트도 같이 만들어 져야 합니다.

laco-dev commented on src/main/kotlin/raicing/model/Car.kt:
```diff
@@ -0,0 +1,5 @@
+package raicing.model
+
+data class Car(val id: Int) {
+    var position: Int = 0
+}
```
자동차는 위치라는 상태를 가지면서 앞으로 전진 또는 멈출 수 있는 자율적인 역할을 수행하는데요.

이러한 관계 속에서 외부에서 자동차의 위치 값을 직접 수정하는건 핵심 로직을 바깥에서 마음대로 수정할 수 있다는 위험한 신호입니다. 🔥 

이 경우 프로퍼티를 내부적으로만 변경 가능하게 만들고 메소드를 통해 내부의 상태를 "스스로' 변경 하게끔 만들어 보면 좋을 것 같아요.

```kotlin
data class Car(val id: Int) {
    var position: Int = 0
          private set  // 외부에서 변경이 불가함!

    fun moveForward() { ... }
}
```

혹은 위치라는 객체를 만들고 해당 객체에 역할을 위임할 수도 있을 것 같아요.
```kotlin
data class Car(val id: Int, private val position: Position = Position.ZERO) {
    val moveForward(...) { ... }
}
```

어떤 방법이 좋을지 고민 해 보시고 그러면 테스트는 어떻게 해 볼 수 있을까? 생각해보세요~

hbm0411 replied:
position을 위한 객체는 over spec 이라고 생각하여 첫번째 방법을 사용했습니다.
private set을 사용하는 방식이 좋아 보이네요!

laco-dev commented on src/main/kotlin/raicing/processor/RaceProcessor.kt:
```diff
@@ -0,0 +1,54 @@
+        return RaceResult(raceResult)
+    }
+
+    private fun goOrStop(car: Car) {
+        val randomCnt = getRandomMoveNumber()
+        if (randomCnt >= BASE_NUM) car.position++
+    }
```
자동차가 랜덤 으로 이동할 수 있도록 만드는 방법은 어떻게 테스트 할 수 있을까요??

자동차는 스스로 움직이는 대신 "움직이는 조건"을 요청해서 협력하도록 만들어 볼 수 있을건데요.
다른 방법도 있을 수 있으니 한번 고민 해 주세요.

hbm0411 replied:
랜덤한 숫자를 받아서 이동하게 하는 함수를 Car 객체로 옮기고, 테스트를 추가하였습니다.

laco-dev commented on src/main/kotlin/raicing/view/OutputView.kt:
```diff
@@ -0,0 +1,21 @@
+import raicing.model.RaceResult
+
+object OutputView {
+    fun resultOutput(raceResultList: List<RaceResult>) {
+        val sb = StringBuilder("========== 결과 ============\n")
+
+        raceResultList.forEach { raceResult ->
+            raceResult.carList.map { car ->
+                sb.append("${car.id}: ")
+                repeat(car.position) {
+                    sb.append("-")
+                }
+                sb.appendLine()
+            }
+            sb.appendLine()
+        }
+        print(sb)
+    }
```
컬렉션과 여러 함수를 복합적으로 사용하면서 함수가 상당히 깊게 만들어졌어요.
함수를 분리해서 보다 가독성을 향상 시켜 보면 어떨까요?

laco-dev replied:
buildString { ... } 이라는 빌더 함수도 있어요!~

hbm0411 replied:
함수 분리 후 buildString을 이용했습니다.

laco-dev reviewed (changes requested):
3단계 미션 잘 해주셨습니다. 💪 

기능 관점의 가장 작은 단위에서 접근 해 보면 처음에는 public 함수로 만들더라도 새로운 기능을 구현 하면서 자연스럽게 private으로 변경이 될 수 있는데요.

이 경우 public 함수가 private 함수를 사용하는 형태로 발전하기 때문에 자연스럽게 테스트가 가능해 집니다.
그래서 만약 정말로 테스트가 불가능한 private 함수가 있다면 해당 역할을 수행하는 다른 객체를 만들 타이밍이 아닐까? 를 고민 해 볼 수 있습니다.

전달드린 피드백을 통해 고민하시고 계신 부분의 가려운 부분을 조금 긁어드릴수 있을 것 같아요.
피드백 확인 후 다시 리뷰 요청 부탁드립니다. 🙏

hbm0411 committed:
refactor: 리뷰 반영

- 사용하지 않는 변수 삭제
- Car의 변수값 외부에서 변경할 수 없도록 수정
- list 변수명 변경(carList -> cars)
- 변수명 명확하게 작성
- list 생성방법 변경

hbm0411 committed:
refactor: 리뷰 반영 - Car, RaceProcessor 구조 변경 및 테스트 코드 작성

hbm0411 committed:
refactor: 결과 출력 view 메소드 분리

laco-dev commented on src/main/kotlin/raicing/model/Car.kt:
```diff
@@ -1,5 +1,26 @@
+            position++
+    }
+
+    fun deepCopy(): Car {
+        val newCar = Car(id)
+        newCar.position = position
+        return newCar
+    }
```
가변적인 값으로 인해 딥카피를 사용하는 목적이라면 position을 불변으로 만드는 전략이 더 효율적으로 보입니다.

내부에서 값이 바뀌었는지 어떻게 변경이 될 것인지 예측하는 것 보다, 하나의 스냅샷 처럼 자동차를 활용 해 보면 어떨까요???

1번 위치에 있는 자동차 인스턴스 A --> 이동 --> 2번 위치에 있는 자동차 인스턴스 B
```kotlin
data class Car(val id: Int, val position: Int) {
    fun moveForward(...) : Car {
        ...
       return copy(position = position + 1)
    }

}
```

laco-dev commented on src/main/kotlin/raicing/model/Car.kt:
```diff
@@ -1,5 +1,26 @@
+    var position: Int = INIT_POSITION
+        private set
+
+    fun moveForwardDependingOption(randomNum: Int) {
+        if (moveForwardOptionCheck(randomNum))
+            position++
+    }
```
네이밍이 외부에서 들어오는 값이 random인지 아닌지는 관심이 없어야 할 것 같아요!

지금은 자동차가 랜덤이라는 비즈니스 로직에 강하게 연결되어 있는데요.
그래서인지 함수명도 내부 구현이 어떤지 관심이 없어야 하는데도 불구하고 내부 구현에 의존하는 형태로 이름이 만들어 졌어요.

자동차가 전진하기 위한 조건을 협력 관게로 구현 해 보면 어떨까요? 🤔 

자동차는 움직일 수 있는지 더이상 스스로 판단하지 않고 "움직여도 되나요?" 라는 관계로 ??
디자인 패턴의 전략 패턴을 참고 해 보세요!

```kotlin
fun moveForward(나 움직여도 되니??) {
   if (나 움직여도 되니?) { 움직였다 }
   else { 움직이지 않았다. }
}
```

hbm0411 replied:
전략 패턴 구현 완료

laco-dev commented on src/main/kotlin/raicing/processor/RaceProcessor.kt:
```diff
@@ -4,42 +4,34 @@ import raicing.model.Car
-        }
-
-        return raceResultList
+    fun initCars(carCount: Int): List<Car> {
+        return List(carCount) { Car(it) }
```
함수를 작은 단위로 분리 해 주신 것 좋습니다. 👍 

그런데 이 함수는 불필요하게(과하게) 분리가 된 것 같아요.
지금 로직으로 보면 정말 필요한 함수인가? 생각이 살짝 듭니다.

제 추측으로는 RaceProcessor를 컨트롤러의 역할로 바라보고 만들어 주신 것 같은데요.
그렇다면 도메인 로직을 수행하는 역할을 분리 해보면 어떨까요? 🤔

laco-dev commented on src/main/kotlin/raicing/processor/RaceProcessor.kt:
```diff
@@ -4,42 +4,34 @@ import raicing.model.Car
-
-            val afterCar = car.copy()
-            afterCar.position = car.position
-            afterCar
+            car.deepCopy()
         }
 
         return RaceResult(raceResult)
     }
 
     private fun goOrStop(car: Car) {
-        val randomCnt = getRandomMoveNumber()
-        if (randomCnt >= BASE_NUM) car.position++
+        val randomNum = getRandomMoveNumber()
+        car.moveForwardDependingOption(randomNum)
```
자동차들의 타입이 List<Car>로 되어 있어 RaceProcessor가 관리를 해 주고 있어요!
그런데 이러한 요구사항이 늘어날때마다 RaceProcessor의 코드가 계속해서 늘어나게 되는 문제로 이어질 수 있는데요.

그러면 결국 private 함수가 많아지면서 어떻게 테스트 하지? 라는 고민에 빠지게 됩니다.

이러한 로직을 "자동차들"이 스스로 해결할 수 있게 자율성을 부여 해 보면 어떨까요?
일급 컬렉션에 대한 개념을 고민 해 보세요
https://jojoldu.tistory.com/412

laco-dev commented on src/test/kotlin/raicing/processor/RaceProcessorTests.kt:
```diff
@@ -0,0 +1,41 @@
+            "5, 3"
+        ]
+    )
+    fun `입력받은 자동차 대수, 시도횟수가 경기 결과 리스트 값의 size와 같아야한다`(carCount: Int, raceCount: Int) {
+        val cars = raceProcessor.initCars(carCount)
+        val result = raceProcessor.race(cars, raceCount)
+
+        assertAll(
+            "carCount",
+            {
+                result.forEach {
+                    assertThat(it.cars.size).isEqualTo(carCount)
+                }
+            }
+        )
+        assertThat(result.size).isEqualTo(raceCount)
+    }
```
테스트 이름을 보고 무슨 뜻이지? 라고 생각을 계속 했네요... 😭 

기댓값과 결과가 직관적으로 드러날 수 있도록 만들어 보면 어떨까요?
```
3대의 자동차로 5번의 경주를 진행하면 5개의 결과와 각 결과의 자동차는 3대이다.
```

laco-dev reviewed (approved):
피드백 반영 잘 해주셨어요! 💯 

구조적인 설계를 어떻게 더 발전시킬지 고민을 많이 하신게 느껴집니다. 🙏 
이런 방법 저런 방법이 있듯이 구조에 대해 발생할 수 있는 단점이 적나라하게 재현이 되는 것 같아서 많이 어렵게 느껴지실 것도 같습니다.

추가로 드린 피드백은 바로 적용하기 보다는 다음 단계를 진행하면서 
"새로운 요구사항을 구현하기 위해서 현재 구조와 제가 제안드린 구조중 어떤게 좋을까?"

라는 관점에서 고민과 비교를 해 보시면 좋겠습니다!

빠른 진행을 위해 3단계 미션은 종료하겠습니다.
다음 미션 화이팅 입니다. 💪
